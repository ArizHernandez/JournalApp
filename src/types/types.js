export const Types = {
  // Auth types
  login :                '[Auth] Login',
  logout:                '[Auth] Logout',
  // Ui types
  uiSetError:            '[UI] Set Error',
  uiRemoveError:         '[UI] Remove Error',
  uiStartLoading:        '[UI] Start Loading',
  uiFinishLoading:       '[UI] Finish Loading',
  // Notes types
  notesAddNew:           '[Notes] New Note',
  notesActive:           '[Notes] Set Active Note',
  notesLoad:             '[Notes] Load Notes',
  notesUpdated:          '[Notes] Update Note',
  notesFileUrl:          '[Notes] Update Image URL',
  notesDeleted:          '[Notes] Delete Note',
  notesLogoutCleaning:   '[Notes] Clear Notes',
}