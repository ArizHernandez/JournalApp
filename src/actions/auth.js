import Swal from 'sweetalert2';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { Types } from '../types/types';
import { notesLogout } from './notes';
import { startLoading, finishLoading, setError } from './ui';


export const startLoginEmailPassword = (email, password) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()))
  };
}

export const startRegisterEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    dispatch(startLoading());

    return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then( async({user}) => {
              await user.updateProfile({displayName: name})

              dispatch(login(user.uid, user.displayName))
            })
            .catch((err) => Swal.fire('Error!', err.message, 'error'))
            .finally(() => dispatch(finishLoading()));
  }
}

export const startGoogleLogin = () => {
  return ((dispatch) => {
    firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => dispatch(login(user.uid, user.displayName)))
  });
}

export const startLogout = () => {
  return async(dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout())
    dispatch(notesLogout());
  }
}

export const login = (uid, displayName) => ({
    type: Types.login,
    payload: {
      uid, 
      displayName
    }
})

export const logout = () => ({
  type: Types.logout
});