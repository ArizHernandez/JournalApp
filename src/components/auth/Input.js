import React, { useState } from 'react'

export const Input = ({required = false, autoComplete ,labelText, placeholder, name, type, value, handleInputChange, expresionRegular, errorMessage}) => {
  const [isValid, setIsValid] = useState("")

  const validate = () => {
    if(expresionRegular){
      if(expresionRegular.test(value)){
        setIsValid(true);
      } else {
        setIsValid(false);
      }
    }
  }

  return (
    <div className="mb-2">
      <label className="mb-1">
        <small className={`
          ${(!isValid && isValid !== "") && "auth__input-invalid-text"}
          ${(isValid && isValid !== "") && "auth__input-valid-text"}
        `}
        >
          {labelText}
        </small>
      </label>

      <div className="auth__input-container">
        <input 
          autoComplete={autoComplete}
          required={required}
          name={name}
          type={type} 
          className={`
            ${(!isValid && isValid !== "") && "auth__input-invalid"} 
            ${(isValid && isValid !=="") && "auth__input-valid"}
            auth__input
          `}
          placeholder={placeholder}
          onKeyUp={validate}
          value={value}
          onChange={handleInputChange}
        />

        {
          (!isValid && isValid !== "")
          ? <i className="auth__input-icon auth__input-invalid-text fa fa-times-circle"></i>
          : (isValid && isValid !== "")
          && <i className="auth__input-icon auth__input-valid-text fa fa-check-circle"></i>
        }
      </div>
      {
        (!isValid && isValid !== "")
        && (<p className="auth__error-message">{errorMessage}</p>)
      }
    </div>
  )
}
