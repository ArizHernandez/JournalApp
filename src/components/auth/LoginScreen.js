import React from 'react'

import validator from 'validator'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'

import { useForm } from '../../hooks/useForm'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { Input } from './Input'

export const LoginScreen = () => {
  const dispatch = useDispatch();
  const {msgError, loading} = useSelector(state => state.ui);
  const userValues = {
    email: '',
    password: ''
  };

  if(process.env.NODE_ENV === 'test'){
    userValues.email = 'test@test.com';
    userValues.password = '123456'
  }

  const [formValues, handleInputChange] = useForm({
    email: userValues.email,
    password: userValues.password
  });
  const {email, password} = formValues;

  const expresiones = {
    email: /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/,
    password: /^.{6,100}$/
  }

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  }
  
  const handleLogin = (e) => {
    e.preventDefault();
    
    if(isFormValid()){
      dispatch(startLoginEmailPassword(email, password))
    }
  }
  
  const isFormValid = () => {

    if(!validator.isEmail(email)){
      dispatch(setError('Email is not valid'));
      return false;
    } else if(password.length < 6){
      dispatch(setError('Password is less to 6 characters'));
      return false;
    }
    
    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className="auth__title txt-center mb-2">Login</h3>
      <form onSubmit={handleLogin}>
        
       <Input
        required={true}
        labelText="Email:"
        name="email"
        type="email"
        placeholder="Email"
        errorMessage="Email is Required"
        expresionRegular={expresiones.email}
        handleInputChange={handleInputChange}
        value={email}
       />
        
       <Input
        labelText="Password:"
        name="password"
        type="password"
        placeholder="Password"
        expresionRegular={expresiones.password}
        handleInputChange={handleInputChange}
        value={password}
       />

        <div className="txt-center mb-2">
          <button
            className="btn btn-primary btn-block"
            type="submit"
            disabled={loading}
          >
            <span>Login</span>
            {
              (loading)
              && (<i className="fa fa-circle-notch fa-spin"></i>)
            }
          </button>
        </div>
      </form>
      
      {
        (msgError)
        && (
          <div className="auth__alert-error">
            <p>{msgError}</p>
          </div>
        )
      }
      
      <hr/>

      <div className="auth__social-network">
        <p>Login with social networks</p>  
        <div 
          className="google-btn mt-2"
          onClick={handleGoogleLogin}
        >
          <div className="google-icon-wrapper">
            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
          </div>
          <p className="btn-text">
            <b>Sign In with Google</b>
          </p>
        </div>
      </div>

      <div className="txt-end">
        <Link 
          to="/auth/register"
          className="link"
        >
          Create new account
        </Link>
      </div>
    </>
  )
}
