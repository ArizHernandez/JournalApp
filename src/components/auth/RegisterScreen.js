import React from 'react'

import validator from 'validator'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';

import { useForm } from '../../hooks/useForm';
import { setError, removeError } from '../../actions/ui';
import { startRegisterEmailPasswordName } from '../../actions/auth';
import { Input } from './Input';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const {msgError, loading} = useSelector( state => state.ui );
  const userValues = {
    name:'',
    email: '',
    password: '',
    password2: ''
  }

  if(process.env.NODE_ENV === 'test'){
    userValues.name = 'test';
    userValues.email = 'test@test.com';
    userValues.password = '123456';
    userValues.password2 = '123456';
  }


  const [formValues, handleInputChange] = useForm({
    name: userValues.name,
    email: userValues.email,
    password: userValues.password,
    password2: userValues.password2
  })
  const {name, email, password, password2} = formValues;

  const expresiones = {
    name: /^[a-zA-ZÀ-ÿ\s]{1,40}$/,
    email: /^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[A-Za-z]+$/,
    password: /^.{6,100}$/
  }

  const handleRegister = (e) => {
    e.preventDefault();

    if(isFormValid()) {
      dispatch(startRegisterEmailPasswordName(email, password, name));
    }
  }

  const isFormValid = () => {
    
    if(name.trim().length === 0){
      dispatch(setError('name is required'))
      return false;
    } else if(!validator.isEmail(email)){
      dispatch(setError('Email is required'));
      return false
    } else if(password !== password2 || password.length < 6){
      dispatch(setError('password is less to 6 characters or is not equal'));
      return false;
    }

    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className="auth__title txt-center mb-2">Register</h3>
      <form onSubmit={handleRegister}>

        <Input
          required={true}
          labelText="Name:"
          name="name"
          type="text"
          placeholder="Name"
          autoComplete="off"
          errorMessage="Name is not valid"
          expresionRegular={expresiones.name}
          handleInputChange={handleInputChange}
          value={name}
        />

        <Input
          required={true}
          labelText="Email:"
          name="email"
          type="email"
          placeholder="example@example.com"
          errorMessage="Email is not valid"
          expresionRegular={expresiones.email}
          handleInputChange={handleInputChange}
          value={email}
        />

        <Input
          required={true}
          labelText="Password:"
          name="password"
          type="password"
          placeholder="******"
          expresionRegular={expresiones.password}
          handleInputChange={handleInputChange}
          value={password}
        />

        <Input
          required={true}
          labelText="Confirm Password:"
          name="password2"
          type="password"  
          placeholder="******"
          expresionRegular={expresiones.password}
          handleInputChange={handleInputChange}
          value={password2}
        />

        <div className="txt-center">
          <button
            className="btn btn-primary btn-block"
            type="submit"
            disabled={loading}
          >
            <span>Sign In </span>
            {
              (loading)
              && (<i className="fa fa-circle-notch fa-spin"></i>)
            }
          </button>
        </div>
      </form>
      <div className="mt-1">
        <span>Alredy registered? </span>
        <Link 
          to="/auth/login"
          className="link"
        >
          logIn
        </Link>
      </div>
      {
        (msgError && !loading) 
        && (
          <div className="auth__alert-error mt-2">
            <p>{msgError}</p>
          </div>
        )
      }
    </>
  )
}
