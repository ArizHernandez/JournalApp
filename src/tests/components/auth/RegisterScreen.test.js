import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { Types } from '../../../types/types';
 
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initilaState = {
  auth: {
    displayName: '',
    uid: ''
  },
  ui: {
    msgError: null, 
    loading: false
  }
}

const store = mockStore(initilaState);

describe('Pruebas en el <RegisterScreen />', () => {
  beforeEach(() => {
    store.clearActions();
  })

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <RegisterScreen />
      </MemoryRouter>
    </Provider>
  );

  test('Debe de cargar correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe de hacer el dispatch de startRegisterEmailPasswordName', () => {
    wrapper.find('form').simulate('submit', {preventDefault: () => {}});
    const actions = store.getActions();

    expect(actions[0]).toEqual({type: Types.uiRemoveError});
  })

  test('Debe de mostrar la caja de alerta correctamente', () => {
    const initilaState = {
      auth: {
        displayName: '',
        uid: ''
      },
      ui: {
        msgError: 'El email es requerido', 
        loading: false
      }
    }
    
    const store = mockStore(initilaState);
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
    expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initilaState.ui.msgError);
  })
  
})
