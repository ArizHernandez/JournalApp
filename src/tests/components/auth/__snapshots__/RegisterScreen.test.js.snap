// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <RegisterScreen /> Debe de cargar correctamente 1`] = `
Array [
  <h3
    className="auth__title txt-center mb-2"
  >
    Register
  </h3>,
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-2"
    >
      <label
        className="mb-1"
      >
        <small
          className="
          false
          
        "
        >
          Name:
        </small>
      </label>
      <div
        className="auth__input-container"
      >
        <input
          autoComplete="off"
          className="
            false 
            
            auth__input
          "
          name="name"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Name"
          required={true}
          type="text"
          value="test"
        />
      </div>
    </div>
    <div
      className="mb-2"
    >
      <label
        className="mb-1"
      >
        <small
          className="
          false
          
        "
        >
          Email:
        </small>
      </label>
      <div
        className="auth__input-container"
      >
        <input
          className="
            false 
            
            auth__input
          "
          name="email"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="example@example.com"
          required={true}
          type="email"
          value="test@test.com"
        />
      </div>
    </div>
    <div
      className="mb-2"
    >
      <label
        className="mb-1"
      >
        <small
          className="
          false
          
        "
        >
          Password:
        </small>
      </label>
      <div
        className="auth__input-container"
      >
        <input
          className="
            false 
            
            auth__input
          "
          name="password"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="******"
          required={true}
          type="password"
          value="123456"
        />
      </div>
    </div>
    <div
      className="mb-2"
    >
      <label
        className="mb-1"
      >
        <small
          className="
          false
          
        "
        >
          Confirm Password:
        </small>
      </label>
      <div
        className="auth__input-container"
      >
        <input
          className="
            false 
            
            auth__input
          "
          name="password2"
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="******"
          required={true}
          type="password"
          value="123456"
        />
      </div>
    </div>
    <div
      className="txt-center"
    >
      <button
        className="btn btn-primary btn-block"
        disabled={false}
        type="submit"
      >
        <span>
          Sign In 
        </span>
      </button>
    </div>
  </form>,
  <div
    className="mt-1"
  >
    <span>
      Alredy registered? 
    </span>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      logIn
    </a>
  </div>,
]
`;
